zkevm-circuits/src/evm_circuit/util/math_gadget.rs

```diff
@@ -783,3 +783,29 @@ impl<F: FieldExt, const N_BYTES: usize> MinMaxGadget<F, N_BYTES> {
         })
     }
 }
+
+// This function generates a Lagrange polynomial in the range [start, end) which
+// will be evaluated to 1 when `exp == value`, otherwise 0
+pub(crate) fn generate_lagrange_base_polynomial<
+    F: FieldExt,
+    Exp: Expr<F>,
+    R: Iterator<Item = usize>,
+>(
+    exp: Exp,
+    val: usize,
+    range: R,
+) -> Expression<F> {
+    let mut numerator = 1u64.expr();
+    let mut denominator = F::from(1);
+    for x in range {
+        if x != val {
+            numerator = numerator * (exp.expr() - x.expr());
```

Comment:
An alternative would be to use one column per target and having boolean selectors. When I was writing this code I wasn't aware of the importance of keeping the expressions' degree as low as possible - and consequently keeping extended domain as small as possible (I was prioritising the number of columns at that time). It seems q_memory_first has a degree of 6, which is not too bad though.

And it depends on the integration with EVM circuit, if it will be integrated in the way that it will use the same extended domain, then it doesn't really matter as EVM circuit will probably have a bigger max poly degree.

So, it probably can stay as it is and use `generate_lagrange_base_polynomial`, but I agree with Edu's comment.

Reply:
> An alternative would be to use one column per target and having boolean selectors.

Thanks for mentioning this!  Actually from a discussion with Han he also brought up this topic.  The thing is that now we have 10 different targets (and we may have more in the future: https://github.com/appliedzkp/zkevm-circuits/pull/335/files#diff-ea6e5fc226ad2891f733fc14acfbe5680b3efc65d96cca49007529272ff82ef3R141).  That will rise the degree even higher.  So Han suggested decomposing the target into 4 bits, and using expressions out those 4 boolean columns to "select" each target.  This way is the middle ground between 1 column with n degree expression and n columns with degree 1 expressions: log2(n) columns with log2(n) degree expression.

Reply:
Cool! Thanks for info!

Reply:
Yes. as far as i remember "decompose flag into bits" tech is also used in mir(polygon zero) <[as far as i remember](https://mirprotocol.org/blog/Fast-recursive-arguments-based-on-Plonk-and-Halo)>